### PR TITLE
Plates

### PR DESCRIPTION
**Pull request by will-moore**
Test sets:

 - https://minio-dev.openmicroscopy.org/idr/idr0033-rohban-pathways/41744_illum_corrected/pr_43/5966.zarr (big plate, 9 fields per well)
 - https://minio-dev.openmicroscopy.org/idr/idr0002-heriche-condensation/plate1_1_013/pr_43/422.zarr (multi-T)
 - https://minio-dev.openmicroscopy.org/idr/idr0004-thorpe-rad52/P101/pr_43/1751.zarr (sparse plate)
 - https://minio-dev.openmicroscopy.org/idr/idr0001-graml-sysgro/JL_120731_S6A/pr_45/2551.zarr (multi-Z, multiple acquisitions)

To test: Open via e.g:
```
$ napari https://minio-dev.openmicroscopy.org/idr/idr0002-heriche-condensation/plate1_1_013/pr_43/422.zarr 
```
 - Plates should be displayed in a grid with Well labels overlaid.
 - You can view an individual Well by appending e.g. `/A/1/` to the URL (actually, this will only work with the last plate above. The others were generated with an acquisition level in the hierarchy and need to add e.g. `0/A/1` or `Run 422/A/1` for idr0002.

Screenshot 2020-10-23 at 07 18 04

### CODE REVIEW

**Comment by will-moore** [comment]
Not doing very well with `pre-commit` here.
With `pre-commit install` I can't commit because flake8 is failing.
When I run it locally it fails on a ton of files that I've not edited.

So I ran flake8 and fixed a bunch of errors, but when I try to commit them
the pre-commit hook fails and `black` undoes all the changes I just made.
So, reverted to `pre-commit uninstall again`!

**Comment by will-moore** [comment]
Exploring ways to add click handling to napari....
With `ome-zarr-py` using this branch.
Save this code to eg. `plate.py`:

```
import numpy as np
import napari
from string import ascii_uppercase


with napari.gui_qt():
    np.random.seed(1)
    viewer = napari.Viewer()

    plate_url = "https://minio-dev.openmicroscopy.org/idr/idr0002-heriche-condensation/plate1_1_013/422_no_T/422.zarr"

    run = "0"
    field = "Field_1"

    plate_layers = viewer.open(plate_url)

    for plate_layer in plate_layers:
        @plate_layer.mouse_drag_callbacks.append
        def get_connected_component_shape(layer, event):

            click_x = layer.coordinates[-1]
            click_y = layer.coordinates[-2]

            plate_width = layer.data.shape[-1]
            plate_height = layer.data.shape[-2]

            # Use 'plate' metadata to work out which Well was clicked
            print(layer.metadata)
            if "plate" in layer.metadata:
                rows = layer.metadata['plate']['rows']
                columns = layer.metadata['plate']['columns']
                print (rows, columns, plate_width, plate_height)
                well_width = plate_width // columns
                well_height = plate_height // rows
                row = int(click_y // well_height)
                column = int(click_x // well_width)
                print('row', row, 'column', column)

                # Open the Image from clicked Well
                field_url = f'{plate_url}/{run}/{ascii_uppercase[row]}/{column + 1}/{field}/'
                print('Open', field_url)
                layers = viewer.open(field_url)

                # Add Well Name to new layers
                well_name = f'{ascii_uppercase[row]}{column + 1}'
                for layer in layers:
                    layer.name = f'{well_name} {layer.name}'
```

Then run `python plate.py`

This loads the stitched plate.
When you click a Well (with one of the stitched plate layers selected), it works out which Well you clicked on and opens that Image - and renames the layers with the Well name. E.g. in this screenshot, A1 was clicked:

Screenshot 2020-10-22 at 22 18 42

**Comment by will-moore** [comment]
That last commit adds Well labels and outlines:

Screenshot 2020-10-23 at 07 18 04

**Comment by joshmoore** [comment]
Nice!

**Comment by will-moore** [comment]
Running `$ pre-commit run -a` locally is passing now...

**Comment by will-moore** [comment]
Travis is failing because `napari 0.4.0` release has dropped python 3.6

**Comment by will-moore** [comment]
With that last commit we can open a single Well, showing fields in a grid:
e.g. 3 x 3 grid stitches together well in this example (but loading stitching together 9 x 1080 * 1080 images is *very* slow
```
$ napari https://minio-dev.openmicroscopy.org/idr/idr0033-rohban-pathways/41744_illum_corrected/pr_43/5966.zarr/0/H/6/
```

Screenshot 2020-11-05 at 15 07 25

**Comment by will-moore** [comment]
Showing a Plate exported with https://github.com/ome/omero-cli-zarr/pull/43

```
napari https://minio-dev.openmicroscopy.org/idr/idr0094-ellinger-sarscov2/screenB/7825.zarr/
```
Screenshot 2020-11-05 at 15 42 18

File: ome_zarr/reader.py [hunk 1]
```diff
@@ -336,6 +343,187 @@ def __init__(self, node: Node) -> None:
...
 
+class Well(Spec):
+    @staticmethod
+    def matches(zarr: BaseZarrLocation) -> bool:
+        return bool("well" in zarr.root_attrs)
+
+    def __init__(self, node: Node) -> None:
+        super().__init__(node)
+        self.well_data = self.lookup("well", {})
+        print("well_data", self.well_data)
```
Comment (joshmoore) [inline ome_zarr/reader.py:354]: These should likely either be removed or be converted to logging statements:

```
napari 101.zarr/
plate_data {'acquisitions': [{'description': 'PlateAcquisition 1 of 2', 'endtime': 1146766431000, 'id': 101, 'maximumfieldcount': None, 'name': 'PlateAcquisition Name 1', 'starttime': 1146766431000}, {'description': 'PlateAcquisition 0 of 2', 'endtime': 1146766431000, 'id': 102, 'maximumfieldcount': None, 'name': 'PlateAcquisition Name 0', 'starttime': 1146766431000}], 'columns': [{'name': '1'}], 'field_count': 4, 'name': 'Plate Name 0', 'rows': [{'name': 'A'}, {'name': 'B'}], 'wells': [{'path': 'B/1'}, {'path': 'A/1'}]}
img_pyramid_shapes [(1, 1, 1, 512, 512), (1, 1, 1, 256, 256), (1, 1, 1, 128, 128), (1, 1, 1, 64, 64)]
target_level 0
creating lazy_reader. level:0 row:0 col:0
lazy_row.shape (1, 1, 1, 512, 512)
creating lazy_reader. level:0 row:1 col:0
lazy_row.shape (1, 1, 1, 512, 512)
lazy_plate (1, 1, 1, 1024, 512)
LOADING tile... A/1/0/0
LOADING tile... B/1/0/0
...data.shape (1, 1, 1, 512, 512)
...data.shape (1, 1, 1, 512, 512)
```

**Review by joshmoore — COMMENTED** [review]
I generated a fake plate:

```
mkfake  plate -runs 2 -rows 2 -fields 2
omero import plate.fake/
omero zarr export Plate:101
napari 101.zarr/
```

and see:

pr-57-a

:+1: (No acquisitions expected yet, right?)

**Comment by will-moore** [comment]
I'm not yet filtering by acquisition here: e.g. 
`napari https://minio-dev.openmicroscopy.org/idr/idr0001-graml-sysgro/JL_120731_S6A/pr_45/2551.zarr/C/4/&acquisition=2661` 
filters by acquisition in vizarr but not in `napari`:

Screenshot 2020-11-16 at 16 17 43

**Review by joshmoore — APPROVED** [review]
Thanks for getting rid of the `print` statements. Locally, `napari -v 101.zarr # <- a plate!` works fine. Getting this in.

**Comment by pwalczysko** [comment]
If I do 

``napari "https://minio-dev.openmicroscopy.org/idr/idr0002-heriche-condensation/plate1_1_013/pr_43/422.zarr/Run 422/A/1"``

then it works and shows the well A1. But, when I play the timelapse, it seems to persistently play the frames between 8  and 18, and loop. Is there any explanation for that ? Update: When I clicked on the timebar, the movie jumped as expected to much higher framenumbers. When I clicked the timelapse play button again, it started to play the whole video, albeit in jumps of some 5 or 10 frames per displayed step, so no smoothness.

**Comment by pwalczysko** [comment]
on the sparse plate

``napari https://minio-dev.openmicroscopy.org/idr/idr0004-thorpe-rad52/P101/pr_43/1751.zarr``


```
18:30:21 ERROR Failed to load 0/D/12/0/3
18:30:21 ERROR Failed to load 0/G/11/0/3
18:30:21 ERROR Failed to load 0/H/10/0/3
18:30:21 ERROR Failed to load 0/F/12/0/3
```

Do we want to supress the message if and when there is no well ?

**Pull request merged**